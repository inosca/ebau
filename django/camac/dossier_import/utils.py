import re
from typing import Any, List, Tuple, Union

import pyexcel_xlsx
from rapidfuzz import fuzz


def get_similar_value(
    original_value: str,
    expected_values: List[str],
    similarity_score_threshold: int = 85,
) -> Union[str, None]:
    """Check if there's a similar expected value."""
    similarity_scores = reversed(
        sorted(
            (
                (expected_value, fuzz.ratio(original_value, expected_value))
                for expected_value in expected_values
            ),
            key=lambda i: i[1],
        )
    )

    return next(
        (
            expected_value
            for expected_value, score in similarity_scores
            if score >= similarity_score_threshold
        ),
        None,
    )


def clean_heading(value: str) -> str:
    """Remove all whitespace characters from headings."""
    return re.sub(r"\s", "", value).strip()


def clean_value(value: Any) -> Any:
    """Remove leading and trailing spaces from string values."""
    return value.strip() if isinstance(value, str) else value


def get_worksheet_headings_and_rows(file) -> Tuple[List[str], List[dict]]:
    """Get headings and rows of an XLSX file."""

    workbook = pyexcel_xlsx.get_data(file)
    worksheet = workbook[list(workbook.keys())[0]]

    headings = [clean_heading(heading) for heading in worksheet[0]]
    rows = [
        dict(zip(headings, [clean_value(cell) for cell in row]))
        for row in worksheet[1:]
    ]

    return headings, rows
