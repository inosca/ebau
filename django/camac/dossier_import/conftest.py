import datetime
import shutil
from collections import OrderedDict
from pathlib import Path

import pytest
from django.core.files import File
from django.core.management import call_command
from django.utils.module_loading import import_string

from camac.dossier_import.tests.test_dossier_import_case import TEST_IMPORT_FILE_PATH
from camac.user.models import Group


@pytest.fixture
def archive_file(settings):
    def get_django_file(name, path=TEST_IMPORT_FILE_PATH, mode="rb"):
        target_path = Path(settings.MEDIA_ROOT) / "dossier_imports"
        target_path.mkdir(parents=True, exist_ok=True)
        shutil.copy(str(Path(path) / name), str(target_path / name))
        return File(open(str(Path(path) / name), mode), name=f"dossier_imports/{name}")

    return get_django_file


@pytest.fixture
def setup_fixtures_required_by_application_config(
    django_db_setup, django_db_blocker, settings
):
    """Set up application configuration data.

    Use this when testing real life dependent procedures and not because you're to
    lazy to setup your test requirements. These fixtures are a bloat and unnecessarily
    slow down the pipeline.
    """

    def load_data(config):
        with django_db_blocker.unblock():
            call_command("loaddata", f"/app/{config}/config/instance.json")
            call_command("loaddata", f"/app/{config}/config/document.json")
            if config == "kt_bern":
                call_command(
                    "loaddata", f"/app/{config}/config/caluma_form_common.json"
                )
                call_command(
                    "loaddata", f"/app/{config}/config/caluma_ebau_number_form.json"
                )
                call_command(
                    "loaddata", f"/app/{config}/config/caluma_dossier_import_form.json"
                )
                call_command(
                    "loaddata", f"/app/{config}/config/caluma_decision_form.json"
                )
            if config == "kt_schwyz":
                call_command("loaddata", f"/app/{config}/config/buildingauthority.json")

    return load_data


@pytest.fixture
def make_workflow_items_for_config(workflow_item_factory):
    def loop_workflow_item_factory(config):
        for pk in {"kt_schwyz": [10, 15]}.get(config, []):
            workflow_item_factory(pk=pk)

    return loop_workflow_item_factory


@pytest.fixture
def dossier_row_sparse():
    return OrderedDict(
        [
            ("ID", "2017-84"),
            ("STATUS", "SUBMITTED"),
            (
                "PROPOSAL",
                "Projekt√§nderung zu Um- und Anbau am Wohnhaus und Solaranlage",
            ),
            ("SUBMIT-DATE", datetime.datetime(2017, 4, 12, 0, 0)),
        ]
    )


@pytest.fixture
def make_dossier_writer(
    db,
    setup_fixtures_required_by_application_config,
    make_workflow_items_for_config,
    settings,
    user,
    group,
    role,
    location,
):
    def init_writer(config):
        setup_fixtures_required_by_application_config(config)
        Group.objects.get_or_create(pk=group.pk, defaults={"role": role})
        writer_cls = import_string(
            settings.APPLICATION["DOSSIER_IMPORT"]["WRITER_CLASS"]
        )
        return writer_cls(
            user_id=user.pk,
            group_id=group.pk,
            location_id=location.pk,
            import_settings=settings.APPLICATION["DOSSIER_IMPORT"],
        )

    return init_writer
