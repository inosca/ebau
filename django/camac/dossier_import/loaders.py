import zipfile
from dataclasses import fields
from enum import Enum
from typing import Generator, Iterable, List, Optional, Tuple

import openpyxl
from django.conf import settings
from django.utils.translation import gettext as _
from pyproj import Transformer

from camac.dossier_import.dossier_classes import (
    Attachment,
    Coordinates,
    Dossier,
    Person,
    PlotData,
)
from camac.dossier_import.messages import (
    LOG_LEVEL_WARNING,
    FieldValidationMessage,
    Message,
    MessageCodes,
)


def numbers(string):
    return int("".join(char for char in str(string) if char.isdigit()) or 0)


def transform_coordinates(n, e, target="epsg:4326"):
    """Transform coordinates from the swiss format to something else."""
    return Transformer.from_crs("epsg:2056", target).transform(n, e)


def safe_join(elements: Iterable, separator=" "):
    """Concatenate elements separated by a given separator.

    Avoid element `None` and allow all types that can be cast to `str`.
    """
    return separator.join(map(str, filter(None, elements))).strip()


class XlsxFileDossierLoader:
    """Load dossiers from xlsx in a zip archive with attachment directories.

    The expected zip file needs to comply with the following:
     - dossiers.xlsx file with column names as defined in this classes.Meta.Columns
     - every directory with a name equal to a dossier's ID will be searched for files
       that then are appended as attachments to the dossier's instance

    Validation and messaging is performed on loading data to the Dossier class.

    The dossier class has simple fields (mandatory and optional) and compound fields (usually
    non-standard objects defined as Dataclass or list thereof).

    Compound fields should provide their loading method that returns the loaded data object plus
    a Message, if validation is required.

    Simple fields should be validiated by the type definition of the Dossier dataclass and be added
    to the Loaders.simple_fields list.

    Any errors are added to the `dossiers._meta.errors` list.

    """

    dossier_class = Dossier
    path_to_dossiers_file: str
    simple_fields = [
        "id",
        "proposal",
        "cantonal_id",
        "street",
        "street_number",
        "city",
        "usage",
        "application_type",
        "submit_date",
        "publication_date",
        "decision_date",
        "construction_start_date",
        "completion_date",
        "decision_date",
        "final_approval_date",
        "profile_approval_date",
        "publication_date",
        "submit_date",
    ]

    required_fields = ("id", "status", "proposal", "submit_date")

    class Column(Enum):
        id = "ID"
        cantonal_id = "CANTONAL-ID"
        status = "STATUS"
        workflow = "WORKFLOW"
        parcel = "PARCEL"
        egrid = "EGRID"
        coordinate_n = "COORDINATE-N"
        coordinate_e = "COORDINATE-E"
        proposal = "PROPOSAL"
        street = "ADDRESS-STREET"
        street_number = "ADDRESS-STREET-NR"
        city = "ADDRESS-CITY"
        usage = "USAGE"
        application_type = "TYPE"
        submit_date = "SUBMIT-DATE"
        publication_date = "PUBLICATION-DATE"
        decision_date = "DECISION-DATE"
        construction_start_date = "CONSTRUCTION-START-DATE"
        profile_approval_date = "PROFILE-APPROVAL-DATE"
        final_approval_date = "FINAL-APPROVAL-DATE"
        completion_date = "COMPLETION-DATE"
        custom_1 = "CUSTOM-1"
        custom_2 = "CUSTOM-2"
        link = "LINK4"
        applicant_first_name = "APPLICANT-FIRST-NAME"
        applicant_last_name = "APPLICANT-LAST-NAME"
        applicant_company = "APPLICANT-COMPANY"
        applicant_street = "APPLICANT-STREET"
        applicant_street_number = "APPLICANT-STREET-NUMBER"
        applicant_zip = "APPLICANT-ZIP"
        applicant_town = "APPLICANT-CITY"
        applicant_phone = "APPLICANT-PHONE"
        applicant_email = "APPLICANT-EMAIL"
        landowner_first_name = "LANDOWNER-FIRST-NAME"
        landowner_last_name = "LANDOWNER-LAST-NAME"
        landowner_company = "LANDOWNER-COMPANY"
        landowner_street = "LANDOWNER-STREET"
        landowner_street_number = "LANDOWNER-STREET-NUMBER"
        landowner_zip = "LANDOWNER-ZIP"
        landowner_town = "LANDOWNER-CITY"
        landowner_phone = "LANDOWNER-PHONE"
        landowner_email = "LANDOWNER-EMAIL"
        projectauthor_first_name = "PROJECTAUTHOR-FIRST-NAME"
        projectauthor_last_name = "PROJECTAUTHOR-LAST-NAME"
        projectauthor_company = "PROJECTAUTHOR-COMPANY"
        projectauthor_street = "PROJECTAUTHOR-STREET"
        projectauthor_street_number = "PROJECTAUTHOR-STREET-NUMBER"
        projectauthor_zip = "PROJECTAUTHOR-ZIP"
        projectauthor_town = "PROJECTAUTHOR-CITY"
        projectauthor_phone = "PROJECTAUTHOR-PHONE"
        projectauthor_email = "PROJECTAUTHOR-EMAIL"

    def load_person(self, dossier_row, prefix):
        """Construct a Person object for a type if any value is given.

        prefix selects respective person properties, e. g. applicant
          `applicant_first_name`
          `applicant_town`
          ...

        """

        person = {
            field.name: dossier_row.get(
                getattr(XlsxFileDossierLoader.Column, f"{prefix}_{field.name}").value
            )
            for field in fields(Person)
        }
        if any(person.values()):
            return [Person(**person)]

    def _load_dossier(self, dossier_row: dict) -> Dossier:
        """Read one line and handle each column.

        Based on the instance of the importer instanciate an ImportCaseData. Loading
        the dossier knows of every column value where to store it.
        """
        dossier = Dossier(
            **{
                key: dossier_row.get(getattr(XlsxFileDossierLoader.Column, key).value)
                for key in self.required_fields
                if key in self.simple_fields
            }
        )
        dossier._meta = Dossier.Meta(
            target_state=dossier_row.get(XlsxFileDossierLoader.Column.status.value),
            workflow=dossier_row.get(XlsxFileDossierLoader.Column.workflow.value)
            or "BUILDINGPERMIT",
        )

        for field in self.required_fields:
            if not dossier_row.get(getattr(XlsxFileDossierLoader.Column, field).value):
                dossier._meta.missing.append(field)

        for key in self.simple_fields:
            setattr(
                dossier,
                key,
                dossier_row.get(getattr(XlsxFileDossierLoader.Column, key).value),
            )

        dossier.plot_data, load_plot_data_errors = self.load_plot_data(dossier_row)
        if load_plot_data_errors:  # pragma: no cover
            dossier._meta.errors += load_plot_data_errors

        dossier.coordinates, load_coordinates_errors = self.load_coordinates(
            dossier_row
        )
        if load_coordinates_errors:  # pragma: no cover
            dossier._meta.errors += load_coordinates_errors

        dossier.applicant = self.load_person(dossier_row, prefix="applicant")
        dossier.landowner = self.load_person(dossier_row, prefix="landowner")
        dossier.project_author = self.load_person(dossier_row, prefix="projectauthor")

        return dossier

    def load_coordinates(
        self, dossier_row
    ) -> Tuple[List[Coordinates], Optional[List[Message]]]:
        out = []
        messages = []
        epoints = dossier_row.get(XlsxFileDossierLoader.Column.coordinate_e.value)
        npoints = dossier_row.get(XlsxFileDossierLoader.Column.coordinate_n.value)
        if not (epoints and npoints):
            return None, messages
        epoints = epoints.split(",") if type(epoints) == str else [epoints]
        npoints = npoints.split(",") if type(npoints) == str else [npoints]
        for e, n in zip(epoints, npoints):
            e, n = numbers(e), numbers(n)
            if not (2480000 < e < 2840000.999) or not (1070000 < n < 1300000.999):
                messages.append(
                    FieldValidationMessage(
                        level=LOG_LEVEL_WARNING,
                        code=MessageCodes.FIELD_VALIDATION_ERROR.value,
                        field="coordinates",
                        detail=_(
                            "The given coordinates (E: %(e)i and N: %(n)i) are not in Switzerland or are not using the Swiss coordinate system (epsg:2056)."
                        )
                        % dict(e=e, n=n),
                    )
                )
                continue

            target_coords = settings.APPLICATION["DOSSIER_IMPORT"].get(
                "TRANSFORM_COORDINATE_SYSTEM"
            )
            if target_coords:
                try:
                    e, n = transform_coordinates(e, n, target_coords)
                except ValueError:  # pragma: no cover
                    messages.append(
                        FieldValidationMessage(
                            level=LOG_LEVEL_WARNING,
                            code=MessageCodes.FIELD_VALIDATION_ERROR.value,
                            field="coordinates",
                            detail=_(
                                "Failed to load and transform coordinates from E: %(e)i and N: %(n)i"
                            )
                            % dict(e=e, n=n),
                        )
                    )
                    continue
            out.append(Coordinates(e=e, n=n))
        return out, messages

    def load_plot_data(
        self, dossier_row
    ) -> Tuple[List[PlotData], Optional[List[Message]]]:
        out = []
        messages = []
        plot_numbers = dossier_row.get(XlsxFileDossierLoader.Column.parcel.value)
        egrids = dossier_row.get(XlsxFileDossierLoader.Column.egrid.value)
        if not (plot_numbers or egrids):
            return None, messages
        try:
            plot_numbers = (
                [p.strip() for p in plot_numbers.split(",")]
                if type(plot_numbers) == str
                else [plot_numbers]
            )

            egrids = (
                [e.strip() for e in egrids.split(",")]
                if type(egrids) == str
                else [egrids]
            )
            municipality = dossier_row.get(
                getattr(XlsxFileDossierLoader.Column, "city").value
            )
            for number, egrid in zip(plot_numbers, egrids):
                out.append(
                    PlotData(
                        number=str(number),
                        egrid=egrid,
                        municipality=municipality,
                    )
                )
        except ValueError:  # pragma: no cover
            messages.append(
                FieldValidationMessage(
                    level=LOG_LEVEL_WARNING,
                    code=MessageCodes.FIELD_VALIDATION_ERROR.value,
                    field="plot-data",
                    detail=_(
                        "Failed to load plot with numbers `%(plot_numbers)s` and egrid values `%(egrids)s`"
                    )
                    % dict(plot_numbers=plot_numbers, egrids=egrids),
                )
            )
        return out, messages

    def load_dossiers(self, path_to_archive: str) -> Generator:
        archive = zipfile.ZipFile(path_to_archive, "r")
        data_file = archive.open("dossiers.xlsx")
        try:
            work_book = openpyxl.load_workbook(data_file, data_only=True)
        except zipfile.BadZipfile:
            raise InvalidImportDataError(
                _("Meta data file in archive is corrupt or not a valid .xlsx file.")
            )
        worksheet = work_book.worksheets[0]
        headings = worksheet[1]
        for row in worksheet.iter_rows(min_row=2):
            dossier = self._load_dossier(
                dict(
                    zip(
                        [heading.value for heading in headings],
                        [cell.value for cell in row],
                    )
                )
            )
            if dossier.id is None:  # pragma: no cover
                continue
            dossier = self._load_attachments(dossier, archive)
            yield dossier

    def _load_attachments(self, dossier, archive):
        for document_name in filter(
            lambda x: x.filename.startswith(f"{dossier.id}/"), archive.infolist()
        ):
            if document_name.filename.endswith("/"):
                continue
            if not dossier.attachments:
                dossier.attachments = []
            dossier.attachments.append(
                Attachment(
                    file_accessor=archive.open(document_name.filename, "r"),
                    name=document_name.filename,
                )
            )
        return dossier


class InvalidImportDataError(Exception):
    pass
