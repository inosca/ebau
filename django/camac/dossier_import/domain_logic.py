import os
import re
import time
from dataclasses import asdict
from functools import wraps
from logging import getLogger

import requests
from caluma.caluma_workflow.models import Case
from django.conf import settings
from django.core.exceptions import ObjectDoesNotExist
from django.utils import timezone
from django.utils.module_loading import import_string
from requests_toolbelt.multipart.encoder import MultipartEncoder

from camac.core.utils import generate_ebau_nr
from camac.document.models import Attachment
from camac.dossier_import.loaders import XlsxFileDossierLoader
from camac.dossier_import.messages import (
    DOSSIER_IMPORT_STATUS_ERROR,
    LOG_LEVEL_ERROR,
    DossierSummary,
    Message,
    MessageCodes,
    update_summary,
)
from camac.dossier_import.models import DossierImport
from camac.instance.models import Instance
from camac.user.models import User
from camac.utils import build_url

logger = getLogger(__name__)


def delay_and_refresh(func):
    @wraps(func)
    def wrapper(dossier_import, *args, **kwargs):
        # django-q is using pickle to transmit arguments. We need to refresh from db, otherwise
        # the task id (which we set immediately after the task is started) is lost when we save the
        # pickled object.
        # We also need to make sure that the view has finished processing (i.e. saved the task id),
        # so we wait shortly.
        time.sleep(0.1)
        dossier_import.refresh_from_db()
        return func(dossier_import, *args, **kwargs)

    return wrapper


@delay_and_refresh
def perform_import(dossier_import, override_config=None):
    try:
        IMPORT_SETTINGS = settings.APPLICATION["DOSSIER_IMPORT"]
        if override_config:
            settings.APPLICATION = settings.APPLICATIONS[override_config]
        configured_writer_cls = import_string(IMPORT_SETTINGS["WRITER_CLASS"])

        loader = XlsxFileDossierLoader()

        writer = configured_writer_cls(
            user_id=User.objects.get(username=IMPORT_SETTINGS["USER"]).pk,
            group_id=dossier_import.group.pk,
            location_id=dossier_import.location and dossier_import.location.pk,
            import_settings=settings.APPLICATION["DOSSIER_IMPORT"],
        )
        dossier_import.messages["import"] = {"details": []}
        for dossier in loader.load_dossiers(dossier_import.source_file.path):
            try:
                message = writer.import_dossier(dossier, str(dossier_import.id))
            except Exception as e:  # pragma: no cover  # noqa: B902
                # We need to catch unhandled exeptions in single dossier imports
                # and keep it going.
                logger.exception(str(e))
                msg = Message(
                    level=LOG_LEVEL_ERROR,
                    code=MessageCodes.UNHANDLED_EXCEPTION.value,
                    detail=str(e),
                )
                message = DossierSummary(
                    dossier_id=dossier.id,
                    status=DOSSIER_IMPORT_STATUS_ERROR,
                    details=[msg],
                )
            dossier_import.messages["import"]["details"].append(asdict(message))
            dossier_import.save()
        update_summary(dossier_import)
        dossier_import.messages["import"]["summary"]["stats"] = {
            "dossiers": Instance.objects.filter(
                **{"case__meta__import-id": str(dossier_import.pk)}
            ).count(),
            "attachments": Attachment.objects.filter(
                **{"instance__case__meta__import-id": str(dossier_import.pk)}
            ).count(),
        }
        dossier_import.messages["import"]["completed"] = timezone.localtime().strftime(
            "%Y-%m-%dT%H:%M:%S%z"
        )
        dossier_import.status = DossierImport.IMPORT_STATUS_IMPORTED

    except Exception as e:  # pragma: no cover  # noqa: B902
        # This is just the last straw. An exception caught here
        # aborts the import session.
        logger.exception(str(e))
        dossier_import.messages["import"]["exception"] = str(e)
        dossier_import.status = DossierImport.IMPORT_STATUS_IMPORT_FAILED

    finally:
        dossier_import.save()

    return dossier_import.status


def get_token():
    DOSSIER_IMPORT = settings.APPLICATION.get("DOSSIER_IMPORT", {})
    r = requests.post(
        DOSSIER_IMPORT.get("PROD_AUTH_URL"),
        {
            "grant_type": "client_credentials",
            "client_id": settings.DOSSIER_IMPORT_CLIENT_ID,
            "client_secret": settings.DOSSIER_IMPORT_CLIENT_SECRET,
        },
    )
    r.raise_for_status()
    return r.json()["access_token"]


@delay_and_refresh
def transmit_import(dossier_import):
    try:
        token = f"Bearer {get_token()}"

        DOSSIER_IMPORT = settings.APPLICATION.get("DOSSIER_IMPORT", {})
        dossier_import.source_file.seek(0)
        fields = {
            "group": str(dossier_import.group.pk),
            "source_file": (
                os.path.basename(dossier_import.source_file.name),
                dossier_import.source_file,
                "application/zip",
            ),
        }
        if DOSSIER_IMPORT.get("LOCATION_REQUIRED", False):
            fields["location_id"] = str(dossier_import.location.pk)

        m = MultipartEncoder(fields=fields)
        r = requests.post(
            build_url(DOSSIER_IMPORT.get("PROD_URL"), "/api/v1/dossier-imports"),
            data=m,
            headers={
                "Content-Type": m.content_type,
                "Authorization": token,
                "x-camac-group": str(DOSSIER_IMPORT.get("PROD_SUPPORT_GROUP_ID")),
            },
        )
        r.raise_for_status()
        dossier_import.status = DossierImport.IMPORT_STATUS_TRANSMITTED

    except Exception as e:  # pragma: no cover # noqa: B902
        logger.exception(e)
        dossier_import.messages["import"]["exception"] = str(e)
        dossier_import.status = DossierImport.IMPORT_STATUS_TRANSMISSION_FAILED

    finally:
        dossier_import.save()
    return dossier_import.status


@delay_and_refresh
def undo_import(dossier_import):
    try:
        # wait shortly to avoid race condition when "in progress" status is saved for
        # very quick "undo" operations
        time.sleep(0.1)
        Instance.objects.filter(
            **{"case__meta__import-id": str(dossier_import.pk)}
        ).delete()
        Case.objects.filter(**{"meta__import-id": str(dossier_import.pk)}).delete()
        dossier_import.delete()
    except Exception as e:  # pragma: no cover # noqa: B902
        logger.exception(e)
        dossier_import.status = DossierImport.IMPORT_STATUS_UNDO_FAILED
        dossier_import.save()


def clean_import(dossier_import):
    try:
        dossier_import.delete_file()
        dossier_import.status = DossierImport.IMPORT_STATUS_CLEANED
    except Exception as e:  # pragma: no cover # noqa: B902
        logger.exception(e)
        dossier_import.messages["import"]["exception"] = str(e)
        dossier_import.status = DossierImport.IMPORT_STATUS_CLEAN_FAILED
    finally:
        dossier_import.save()
    return dossier_import.status


def get_or_create_ebau_nr(ebau_number, service, submit_date=None):
    """Validate a proposed ebau-number to match its service domain or get a new one."""
    pattern = re.compile("([0-9]{4}-[1-9][0-9]*)")
    result = pattern.search(str(ebau_number))
    if result:
        try:
            match = result.groups()[0]
            case = Case.objects.filter(**{"meta__ebau-number": match}).first()
            if case.instance.services.filter(service_id=service.pk).exists():
                return match
        except AttributeError:
            pass

    return generate_ebau_nr(submit_date.year) if submit_date else None


def set_status_callback(task):  # pragma: no cover
    # this is not covered as long as the sync mode of django-q is not fixed
    dossier_import = task.args[0]
    try:
        dossier_import.refresh_from_db()
    except ObjectDoesNotExist:
        # the undo task deletes the instance on success:
        return

    dossier_import.status = task.result or dossier_import.set_progressing_to_failed()
    dossier_import.save()
