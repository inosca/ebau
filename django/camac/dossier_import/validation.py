import datetime
from enum import Enum
import zipfile
from typing import List

import openpyxl
from django.utils import timezone
from django.utils.translation import gettext as _
from rest_framework.exceptions import ValidationError

from camac.dossier_import import messages
from camac.dossier_import.loaders import InvalidImportDataError
from camac.dossier_import.models import DossierImport

from .config.common import mimetypes
from .dossier_classes import Dossier
from .messages import MessageCodes
from .loaders import XlsxFileDossierLoader


class TargetStatus(Enum):
    SUBMITTED = "SUBMITTED"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
    WRITTEN_OFF = "WRITTEN OFF"
    DONE = "DONE"


def verify_source_file(source_file: str) -> str:
    """
    Verify source file is valid archive and has content.

    Verfication here goes pretty far. But since it's about rather big archive files
    we do not want to store any file that is not going to survive the first
    step of processing anyway.

    Failing to validate source_file results in an APIError i. e. bad request

    The following steps are required:
     - archive must be a ZIP file
     - the archive must contain a dossiers.xlsx
     - the dossiers.xlsx must in fact be a XLSX file
    """
    if source_file is None:
        raise ValidationError(_("To start an import please upload a file."))

    try:
        archive = zipfile.ZipFile(source_file)
    except zipfile.BadZipfile:
        raise ValidationError(_("Uploaded file is not a valid .zip file"))
    try:
        metadata = archive.open("dossiers.xlsx")
    except KeyError:
        raise ValidationError(
            _("No metadata file 'dossiers.xlsx' found in uploaded archive.")
        )
    try:
        openpyxl.load_workbook(metadata, data_only=True)
    except zipfile.BadZipfile:
        raise ValidationError(
            _("Metadata file `dossiers.xlsx` is not a valid .xlsx file.")
        )
    return source_file


def validate_attachments(archive: zipfile.ZipFile, dossier_ids: List[str]):
    dirs = {
        zipinfo.filename.split("/")[0]
        for zipinfo in archive.filelist
        if zipinfo.filename.endswith("/") and len(zipinfo.filename.split("/")) < 3
    }
    orphan_dirs = sorted(list(dirs - set(dossier_ids)))
    result = []
    if orphan_dirs:
        result.append(
            _(
                "%(count)i document folders were not found in the metadata file and will not be imported:\n%(entries)s"
            )
            % dict(count=len(orphan_dirs), entries=", ".join(orphan_dirs))
        )
    dossiers_without_documents = list(set(dossier_ids) - dirs)
    if dossiers_without_documents:
        result.append(
            _("%(count)i dossiers have no document folder.")
            % dict(count=len(dossiers_without_documents))
        )
    return result


def get_attachment_validation_stats(archive: zipfile.ZipFile, dossier_ids: List[str]):
    return sum(
        [
            1
            for url in archive.filelist
            if url.filename.split("/")[0] in dossier_ids
            and not url.filename.endswith("/")
            and mimetypes.guess_type(url.filename)
        ]
    )


# flake8: noqa: C901
def validate_zip_archive_structure(instance_pk, clean_on_fail=True) -> DossierImport:
    """
    ZIP archive validation.

    scans the archive and best guesses the outcome of actually importing it.
    """
    dossier_import = DossierImport.objects.get(pk=instance_pk)

    archive = zipfile.ZipFile(dossier_import.source_file.path, "r")
    data_file = archive.open("dossiers.xlsx")
    try:
        work_book = openpyxl.load_workbook(data_file, data_only=True)
    except zipfile.BadZipfile:
        raise InvalidImportDataError(
            _("Meta data file in archive is corrupt or not a valid .xlsx file.")
        )
    worksheet = work_book.worksheets[0]
    headings = [h for h in worksheet[1] if h.value is not None]

    required_columns = ["ID", "STATUS", "PROPOSAL", "SUBMIT-DATE"]
    heading_values = [col.value for col in headings]
    missing = set(required_columns) - set(heading_values)
    if missing:
        raise InvalidImportDataError(
            _(
                "Meta data file in archive is missing required columns %(missing)s. Found %(found)s"
            )
            % dict(missing=missing, found=heading_values)
        )
    extra = set(heading_values) - set([e.value for e in XlsxFileDossierLoader.Column])
    if extra:
        dossier_import.messages["validation"]["summary"]["warning"].append(
            _(
                "Found unknown columns: %(extra)s. Those columns will be ignored while importing."
            )
            % dict(extra=extra)
        )

    status_column = next((col for col in headings if col.value == "STATUS"))
    submit_date_column = next((col for col in headings if col.value == "SUBMIT-DATE"))
    id_column = next((col for col in headings if col.value == "ID"))

    dossier_msgs = []
    deleted_rows = 0
    for cell in worksheet[id_column.column_letter]:
        if cell.value is not None:
            continue
        worksheet.delete_rows(cell.row, 1)
        deleted_rows += 1

    dossier_ids = [cell.value for cell in worksheet[id_column.column_letter][1:]]

    dupes = set([d for d in dossier_ids if dossier_ids.count(d) > 1])

    for dupe in dupes:
        messages.append_or_update_dossier_message(
            dupe,
            "id",
            f"{dupe} is not unique.",
            MessageCodes.DUPLICATE_IDENTFIER_ERROR.value,
            dossier_msgs,
        )

    dossiers_success = list(set(dossier_ids))

    for date_column in [col for col in headings if col.value.endswith("-DATE")]:
        for cell in worksheet[date_column.column_letter][1:]:
            dossier_id = worksheet[f"A{cell.row}"].value
            field_name = worksheet[f"{cell.column_letter}1"].value.lower()
            try:
                assert cell.value is None or type(cell.value) == datetime.datetime
            except AssertionError:
                messages.append_or_update_dossier_message(
                    dossier_id,
                    field_name,
                    str(cell.value),
                    MessageCodes.DATE_FIELD_VALIDATION_ERROR.value,
                    dossier_msgs,
                )
                try:
                    dossiers_success.remove(dossier_id)
                except ValueError:
                    pass
                continue

    if status_column:
        for cell in worksheet[status_column.column_letter][1:]:
            dossier_id = worksheet[f"A{cell.row}"].value
            if cell.value not in [e.value for e in TargetStatus]:
                if cell.value is None:
                    messages.append_or_update_dossier_message(
                        dossier_id,
                        "status",
                        None,
                        MessageCodes.MISSING_REQUIRED_VALUE_ERROR.value,
                        dossier_msgs,
                        level=messages.LOG_LEVEL_ERROR,
                    )
                    try:
                        dossiers_success.remove(dossier_id)
                    except ValueError:  # pragma: no cover
                        pass
                    continue

                messages.append_or_update_dossier_message(
                    dossier_id,
                    "status",
                    cell.value,
                    MessageCodes.STATUS_CHOICE_VALIDATION_ERROR.value,
                    dossier_msgs,
                    level=messages.LOG_LEVEL_ERROR,
                )
                try:
                    dossiers_success.remove(dossier_id)
                except ValueError:  # pragma: no cover
                    pass

    if submit_date_column:
        for cell in worksheet[submit_date_column.column_letter][1:]:
            dossier_id = worksheet[f"A{cell.row}"].value
            if not cell.value:
                messages.append_or_update_dossier_message(
                    dossier_id,
                    "submit_date",
                    None,
                    MessageCodes.MISSING_REQUIRED_VALUE_ERROR.value,
                    dossier_msgs,
                    level=messages.LOG_LEVEL_ERROR,
                )
                try:
                    dossiers_success.remove(dossier_id)
                except ValueError:  # pragma: no cover
                    pass

    for msg in dossier_msgs:
        messages.update_messages_section_detail(
            msg, dossier_import, section="validation"
        )

    dossier_import = messages.update_summary(dossier_import)
    dossier_import.messages["validation"]["summary"]["warning"] += validate_attachments(
        archive, dossier_ids
    )
    dossier_import.messages["validation"]["summary"]["stats"] = {
        "attachments": get_attachment_validation_stats(archive, dossier_ids),
        "dossiers": len(dossiers_success),
    }

    dossier_import.messages["validation"]["completed"] = timezone.localtime().strftime(
        "%Y-%m-%dT%H:%M:%S%z"
    )
    dossier_import.save()

    dossier_import.status = dossier_import.IMPORT_STATUS_VALIDATION_SUCCESSFUL
    if dossier_import.messages["validation"]["summary"]["error"]:
        dossier_import.status = dossier_import.IMPORT_STATUS_VALIDATION_FAILED
        if clean_on_fail:
            dossier_import.source_file.delete()
    dossier_import.save()

    return dossier_import
