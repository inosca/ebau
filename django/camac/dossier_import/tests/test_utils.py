import pytest

from camac.dossier_import.loaders import XlsxFileDossierLoader
from camac.dossier_import.utils import get_similar_value
from camac.dossier_import.validation import TargetStatus, validate_extra_columns

existing_columns = [e.value for e in XlsxFileDossierLoader.Column]
existing_status = [e.value for e in TargetStatus]


@pytest.mark.parametrize(
    "input,existing,output",
    [
        ("THIS_REALLY_DOESNT_EXIST", existing_columns, None),
        ("PUBLICATIONDATE", existing_columns, "PUBLICATION-DATE"),
        ("CONSTRUCTION_START_DATE", existing_columns, "CONSTRUCTION-START-DATE"),
        ("APLICANT-FIRSTNAME", existing_columns, "APPLICANT-FIRST-NAME"),
        ("WEIRD_STATUS", existing_status, None),
        ("SUBIMTTED", existing_status, "SUBMITTED"),
    ],
)
def test_similar_value(input, existing, output):
    assert get_similar_value(input, existing) == output


def test_validate_extra_columns(snapshot):
    assert (
        validate_extra_columns(
            [
                "APPLICANT-FIRST-NAME",
                "SOMEOTHERCOLUMN",
                "PUBLICATIONDATE",
                "SUBIMT-DATE",
            ]
        )
        == snapshot
    )
