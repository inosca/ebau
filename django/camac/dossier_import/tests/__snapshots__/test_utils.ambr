# serializer version: 1
# name: test_validate_extra_columns
  '''
  Die folgenden Spalten wurden nicht erkannt und werden beim Import ignoriert:
  PUBLICATIONDATE - meinten Sie "PUBLICATION-DATE"?
  SOMEOTHERCOLUMN
  SUBIMT-DATE - meinten Sie "SUBMIT-DATE"?.
  '''
# ---
